### ISSUE TITLE
API :: Private pages not listed when authenticated

### ISSUE THREAD

**Issue by spoetnik** [wagtail/wagtail#3672]
### Issue Summary

I have set some pages, with childpages, to private. Ofter logging in on the website, these pages are shown. When browsing the /pages endpoint in the API, those private pages are not listed. The DRF interface shows my logged in username, so I am authenticated.

### Steps to Reproduce
1) Create a group called 'Private'
2) Create a user called 'Private', which is member of the 'Private' group.
3) Create a page, add subpages if you feel like it, and set the privacy to 'Private, accessible to users in specific groups', and select the group 'Private'. Save
4) In the Live-view, go to the created page with the Privacy settings. You are redirected to a login page.
5) Login with the user 'Private'
6) The page is available to you, and can be seen.
7) Go to the /pages API endpoint, /api/v1/pages/
8) Find the page with the Privacy settings. Bang head to the wall...
9) Notice in the upper right corner you are authenticated as user 'Private'

I would expect the 'Private' page to be available in the public API after authentication. Only Public pages are shown.

### Technical details

* Python version: Python 3.6.1
* Django version: Django==1.11.2
* Wagtail version: wagtail==1.10.1

**Comment by gasman** [comment]
Thanks for the report @spoetnik! I can confirm that the API doesn't currently support authentication to access private pages - it just filters to public/live pages:

https://github.com/wagtail/wagtail/blob/deeee4a5978be50eb6f12132ca1f00d8cdc19f05/wagtail/api/v2/endpoints.py#L402-L403

**Comment by jonathanfmachado** [comment]
Hi @spoetnik. Did you manage to find a way around this issue?

> I would expect the 'Private' page to be available in the public API after authentication. Only Public pages are shown.

**Comment by peteyb** [comment]
Hi @jonathanfmachado 

Not sure if this is of help to you but a relatively simple solution is to subclass the PagesAPIEndpoint class and override the get_queryset() method to check if the user is authenticated.

It does not take into account the various private options, such as being part of a group, but this may suffice for your purposes.

```
from wagtail.api.v2.endpoints import PagesAPIEndpoint
from wagtail.api.v2.utils import (BadRequestError, filter_page_type,
                                  page_models_from_string)
from wagtail.core.models import Page


class MyPagesAPIEndpoint(PagesAPIEndpoint):
    """
    Override stock wagtail pages api to handle private pages
    """
    def get_queryset(self):
        request = self.request

        # Allow pages to be filtered to a specific type
        try:
            models = page_models_from_string(request.GET.get('type', 'wagtailcore.Page'))
        except (LookupError, ValueError):
            raise BadRequestError("type doesn't exist")

        if not models:
            models = [Page]

        if len(models) == 1:
            queryset = models[0].objects.all()
        else:
            queryset = Page.objects.all()

            # Filter pages by specified models
            queryset = filter_page_type(queryset, models)

        if self.request.user.is_authenticated:
            # Get all live pages
            queryset = queryset.live()
        else:
            # Get live pages that are not in a private section
            queryset = queryset.public().live()

        # Filter by site
        queryset = queryset.descendant_of(request.site.root_page, inclusive=True)

        return queryset
```

**Comment by jonathanmach** [comment]
> Hi @jonathanfmachado
> 
> Not sure if this is of help to you but a relatively simple solution is to subclass the PagesAPIEndpoint class and override the get_queryset() method to check if the user is authenticated.
> 
> It does not take into account the various private options, such as being part of a group. It is also not future proof from changes to that method in the core wagtail project but may suffice for your purposes
> 
> ```
> from wagtail.api.v2.endpoints import PagesAPIEndpoint
> from wagtail.api.v2.utils import (BadRequestError, filter_page_type,
>                                   page_models_from_string)
> from wagtail.core.models import Page
> 
> 
> class MyPagesAPIEndpoint(PagesAPIEndpoint):
>     """
>     Override stock wagtail pages api to handle private pages
>     """
>     def get_queryset(self):
>         request = self.request
> 
>         # Allow pages to be filtered to a specific type
>         try:
>             models = page_models_from_string(request.GET.get('type', 'wagtailcore.Page'))
>         except (LookupError, ValueError):
>             raise BadRequestError("type doesn't exist")
> 
>         if not models:
>             models = [Page]
> 
>         if len(models) == 1:
>             queryset = models[0].objects.all()
>         else:
>             queryset = Page.objects.all()
> 
>             # Filter pages by specified models
>             queryset = filter_page_type(queryset, models)
> 
>         if self.request.user.is_authenticated:
>             # Get all live pages
>             queryset = queryset.live()
>         else:
>             # Get live pages that are not in a private section
>             queryset = queryset.public().live()
> 
>         # Filter by site
>         queryset = queryset.descendant_of(request.site.root_page, inclusive=True)
> 
>         return queryset
> ```
> 
> ```
> # api.py
> 
> api_router.register_endpoint('pages', MyPagesAPIEndpoint)
> ```

Hi! Thank you very much for the response!

**Comment by diegoleft** [comment]
Hello guys,

Is there a way to get the "privacy" meta in the endpoint?
I would like to know if the list_view can give me that info to restrict in the detail_view but not in the list.

Thanks in advance

**Comment by zerolab** [comment]
This was fixed in #7385 and available in 2.16+
